### PR TITLE
CLI support for listing out current app resources

### LINKED ISSUE #1136: I should be able to see a list of resources on my app from the CLI.

### PR DESCRIPTION
**Pull request by patcoll**
With corresponding provider names. Fixes #1136

### CODE REVIEW

**Comment by titanous** [comment]
Thanks! The commit message needs a `cli: ` prefix.

File: cli/resource.go [hunk 1]
```diff
@@ -24,7 +26,24 @@ func runResource(args *docopt.Args, client *controller.Client) error {
...
+	if err != nil {
+		return err
+	}
+
+	w := tabWriter()
+	defer w.Flush()
+
+	var provider *ct.Provider
+
+	listRec(w, "ID", "ProviderID", "ProviderName")
```
Comment (titanous) [inline cli/resource.go]: I think we should add spaces between Provider and ID/Name, so `Provider ID`, etc.

**Comment by patcoll** [comment]
Should I amend it to be this then? `cli:Add support for listing out current app resources, with corresponding provider names.`

**Comment by titanous** [comment]
The first line should be `cli: Add provider list command`, you can describe in more detail starting on the third line. The commit should be amended and force-pushed to the same branch.

File: cli/resource.go [hunk 2]
```diff
@@ -24,7 +26,24 @@ func runResource(args *docopt.Args, client *controller.Client) error {
...
+	}
+
+	w := tabWriter()
+	defer w.Flush()
+
+	var provider *ct.Provider
+
+	listRec(w, "ID", "Provider ID", "Provider Name")
+	for _, j := range resources {
+		provider, err = client.GetProvider(j.ProviderID)
```
Comment (titanous) [inline cli/resource.go]: This error should be handled otherwise the next line could trigger a nil panic.

Comment (patcoll) [inline cli/resource.go]: Right indeed. Added a clause to continue if error found.

Comment (titanous) [inline cli/resource.go]: The new code has the potential to ignore errors (error, then success would hide the issue). I'd suggest just returning the error right away.

Comment (patcoll) [inline cli/resource.go]: Should CLI commands halt on first error? Asking earnestly. I wonder if it's
better to output as much as you can in a streaming format. Perhaps tabular
output to stdout, errors to stderr.

On Monday, March 2, 2015, Jonathan Rudenberg <notifications@github.com>
wrote:

> In cli/resource.go
> <https://github.com/flynn/flynn/pull/1137#discussion_r25658959>:
>
> > @@ -24,7 +26,24 @@ func runResource(args *docopt.Args, client *controller.Client) error {
> >  	if args.Bool["add"] {
> >  		return runResourceAdd(args, client)
> >  	}
> > -	return fmt.Errorf("Top-level command not implemented.")
> > +
> > +	resources, err := client.AppResourceList(mustApp())
> > +	if err != nil {
> > +		return err
> > +	}
> > +
> > +	w := tabWriter()
> > +	defer w.Flush()
> > +
> > +	var provider *ct.Provider
> > +
> > +	listRec(w, "ID", "Provider ID", "Provider Name")
> > +	for _, j := range resources {
> > +		provider, err = client.GetProvider(j.ProviderID)
>
> The new code has the potential to ignore errors (error, then success would
> hide the issue). I'd suggest just returning the error right away.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/flynn/flynn/pull/1137/files#r25658959>.
>


-- 
Patrick Collins

Comment (titanous) [inline cli/resource.go]: > Should CLI commands halt on first error?

Until we add retry support, I think that makes sense.

> Perhaps tabular output to stdout, errors to stderr.

I'm pretty sure we already do this.

Comment (patcoll) [inline cli/resource.go]: Fixed.

**Comment by titanous** [comment]
Looks good! Please squash the last commit and rebase on the latest master and I'll merge.

**Comment by patcoll** [comment]
Done and ready for merge.

**Comment by titanous** [comment]
Thanks!

**Pull request merged**